wiki/Client Security.md

```diff
@@ -0,0 +1,95 @@
+## Client Security
+
+As noted in https://github.com/polymorpher/one-wallet/issues/5, in ONE Wallet v0.1 the authenticator cannot provide security protection if the client is compromised. A compromised client means the information stored on the client is leaked to a malicious third-party. In the current setup, the third-party may use these leaked information to easily reverse engineer (by brute force) the proof needed to perform an unauthorized transfer from ONE Wallet. The 6-digit code input from Google Authenticator would not be required.
+
+To quote from the Github issue:
+
+> The core weakness resides in the small search space in how EOTP is generated from an OTP and several other client-side parameters (hseed, nonce). See code at
+> 
+> https://github.com/polymorpher/one-wallet/blob/ae1b113cdfe022d2ec65152e3b8654f27f756d99/code/lib/onewallet.js#L25
+> 
+> Since the OTP is confined to a 6-digit numerical number, and the generation process relies on SHA256, the right OTP (and by extension, its EOTP) at any given time can be easily enumerated by brute-force, for any given leaf hash value. See for the code of doing so at 
+> 
+> https://github.com/polymorpher/one-wallet/blob/afe39e657df1522ee8941e386025c4ddafc3ab94/code/lib/onewallet.js#L141 
+
+An adequate solution must make it exceedingly difficult, if not practically impossible, for any party to perform any transfer (or similar operations) without the input from the Google Authenticator. At the same time, it must not significantly slow down the wallet creation process, or the process of making an authorized operation with the correct authenticator code.
+
+This leaves us several options to enhance the security:
+
+1. Substantially increase the search space required for brute force attack, from the current 1 million possible OTPs.
```

Comment:
By @ivan-homoliak-sutd :
```
I do not think this will help since the attacker who already has the seed (i.e., its hash) can bruteforce just one particular EOTP in the near future. She does not need to make bruteforce across multiple EOTPs to significantly increase her chances of success. Maybe you mean the scrambled design, but even here, attacker might do some tricks to get EOTPs (see below)
```

Reply:
I should clarify in the article that the search space here is within the context of one EOTP. To brute-force, the attacker needs to enumerate all possible inputs to and check which one generates the target EOTP. Since the attacker is assumed to have already obtained all information at the client, the input space is left with only 1 million possibilities: the OTP from 0 to 100000, which the attacker needs to enumerate

Reply:
So you do not assume that the attacker might also try to brute-force other EOTPs, whose hashes (or hashes of their near parents) she might obtain from the blockchain? I do not see the reason why it should not be considered. Can you outline?

Reply:
They can brute-force other EOTPs but they can only obtain past EOTPs - which don't seem to be useful to brute-force?

Reply:
Ok, got it.